File: Weather/Weather/Services/NWSService.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

namespace Weather.Services
{
    class NWSService
    {
        static readonly string baseAddress = "https://api.weather.gov";
        readonly HttpClient client;
        DateTime lastRefresh = DateTime.MinValue;
        void NWSService_LocationChanged(object sender, EventArgs e) => UpdateLocation();

        public bool IsValid => Location != null;

        public event EventHandler DataUpdated;

        public double QueryLatitude { private set; get; }
        public double QueryLongitude { private set; get; }

        public string Location { private set; get; }

        // current conditions
        public double? Temperature { private set; get; }
        public double? WindChill { private set; get; }
        public double? HeatIndex { private set; get; }
        public DateTime? Timestamp { private set; get; }
        public string TextDescription { private set; get; }
        public double? WindDirection { private set; get; }
        public double? WindSpeed { private set; get; }
        public double? DewPoint { private set; get; }
        public double? RelativeHumidity { private set; get; }
        public string ConditionIconUrl { private set; get; }
        public double? Visibility { private set; get; }
        public double? WindGust { private set; get; }
        public double? Pressure { private set; get; }

        // forecast
        public List<string> ForecastLabels { private set; get; }
        public List<string> ForecastIcons { private set; get; }
        public List<string> ForecastDescriptions { private set; get; }
        public List<double> ForecastTemperatures { private set; get; }
        public List<string> WordedForecast { private set; get; }

        static NWSService service;
        static readonly object serviceLock = new object();
        public static NWSService Service
        {
            get
            {
                lock (serviceLock)
                    if (service == null)
                        service = new NWSService();
                return service;
            }
        }

        // station
        string StationId;

        // grid point
        string GridWFO;
        int? GridX;
        int? GridY;

        NWSService()
        {
            client = new HttpClient();
            client.DefaultRequestHeaders.Add("User-Agent", "Weather app");

            UpdateLocation();
            LocationService.Service.LocationChanged += NWSService_LocationChanged;
        }

        private void UpdateLocation()
        {
            var location = LocationService.Service.Location;

            if (QueryLatitude != location.Latitude || QueryLongitude != location.Longitude)
            {
                QueryLatitude = location.Latitude;
                QueryLongitude = location.Longitude;

                StationId = null;
                GridWFO = null;
                GridX = null;
                GridY = null;

                Location = null;
                Temperature = null;
                WindChill = null;
                HeatIndex = null;
                Timestamp = null;
                TextDescription = null;
                WindDirection = null;
                WindSpeed = null;
                DewPoint = null;
                RelativeHumidity = null;
                ConditionIconUrl = null;
                ForecastLabels = null;
                ForecastIcons = null;
                ForecastDescriptions = null;
                ForecastTemperatures = null;
                WordedForecast = null;

                lastRefresh = DateTime.MinValue;

                Refresh();
            }
        }

        readonly SemaphoreSlim refreshLock = new SemaphoreSlim(1);
        public async void Refresh()
        {
            await refreshLock.WaitAsync();

            if (!IsValid || DateTime.UtcNow - lastRefresh > TimeSpan.FromMinutes(15))
            {
                try
                {
                    await GetCurrentConditions();
                    await GetForecast();

                    lastRefresh = DateTime.UtcNow;
                }
                catch (Exception ex)
                {
                    System.Diagnostics.Debug.WriteLine(ex);
                }
            }

            refreshLock.Release();

            if (DataUpdated != null)
                DataUpdated.Invoke(this, EventArgs.Empty);
        }

    async Task GetStationFromLocation()
        {
            var uri = new Uri($"{baseAddress}/points/{QueryLatitude},{QueryLongitude}/stations");
            var response = await client.GetAsync(uri);
            if (response.IsSuccessStatusCode)
            {
                // extract the needed data from the json string
                var content = await response.Content.ReadAsStringAsync();
                var jObject = JObject.Parse(content);
                Location = (string)jObject["features"][0]["properties"]["name"];
                StationId = (string)jObject["features"][0]["properties"]["stationIdentifier"];
            }
        }

        async Task GetCurrentConditions()
        {
            if (StationId == null)
                await GetStationFromLocation();

            // make network request
            var uri = new Uri($"{baseAddress}/stations/{StationId}/observations/latest?require_qc=false");
            var response = await client.GetAsync(uri);
            if (response.IsSuccessStatusCode)
            {
                // extract the needed data from the json string
                var content = await response.Content.ReadAsStringAsync();
                var jObject = JObject.Parse(content);

                var timestampString = (string)jObject["properties"]["timestamp"];
                Timestamp = DateTime.Parse(timestampString);

                double.TryParse((string)jObject["properties"]["dewpoint"]["value"], out double dewpoint);
                double.TryParse((string)jObject["properties"]["relativeHumidity"]["value"], out double relativeHumidity);
                double.TryParse((string)jObject["properties"]["temperature"]["value"], out double temperature);
                double.TryParse((string)jObject["properties"]["windDirection"]["value"], out double windDirection);
                double.TryParse((string)jObject["properties"]["windSpeed"]["value"], out double windSpeed);
                if (double.TryParse((string)jObject["properties"]["windGust"]["value"], out double windGust))
                    WindGust = MetersPerSecondToMPH(windGust);
                if (double.TryParse((string)jObject["properties"]["barometricPressure"]["value"], out double barometricPressure))
                    Pressure = PaToHG(barometricPressure);
                if (double.TryParse((string)jObject["properties"]["visibility"]["value"], out double visibility))
                    Visibility = MToMiles(visibility);
                if (double.TryParse((string)jObject["properties"]["windChill"]["value"], out double windChill))
                    WindChill = CToF(windChill);
                if (double.TryParse((string)jObject["properties"]["heatIndex"]["value"], out double heatIndex))
                    HeatIndex = CToF(heatIndex);

                DewPoint = CToF(dewpoint);
                RelativeHumidity = relativeHumidity;
                Temperature = CToF(temperature);
                WindDirection = windDirection;
                WindSpeed = MetersPerSecondToMPH(windSpeed);

                TextDescription = (string)jObject["properties"]["textDescription"];
                ConditionIconUrl = (string)jObject["properties"]["icon"];
            }
        }

        double CToF(double celcius) => 9.0 / 5.0 * celcius + 32.0;
        double MetersPerSecondToMPH(double metersPerSecond) => metersPerSecond * 2.237;
        double PaToHG(double pascals) => pascals / 3386.0;
        double MToMiles(double meters) => meters / 1609.0;

        async Task GetGridPoint()
        {
            var uri = new Uri($"{baseAddress}/points/{QueryLatitude},{QueryLongitude}");
            var response = await client.GetAsync(uri);
            if (response.IsSuccessStatusCode)
            {
                // extract the needed data from the json string
                var content = await response.Content.ReadAsStringAsync();
                var jObject = JObject.Parse(content);
                GridWFO = (string)jObject["properties"]["cwa"];
                GridX = (int)jObject["properties"]["gridX"];
                GridY = (int)jObject["properties"]["gridY"];
            }
        }

        async Task GetForecast()
        {
            if (!GridX.HasValue || !GridY.HasValue || string.IsNullOrWhiteSpace(GridWFO))
                await GetGridPoint();

            // make network request
            var uri = new Uri($"{baseAddress}/gridpoints/{GridWFO}/{GridX},{GridY}/forecast");
            var response = await client.GetAsync(uri);
            if (response.IsSuccessStatusCode)
            {
                // extract the needed data from the json string
                var content = await response.Content.ReadAsStringAsync();
                var jObject = JObject.Parse(content);

                ForecastLabels = new List<string>();
                ForecastIcons = new List<string>();
                ForecastDescriptions = new List<string>();
                ForecastTemperatures = new List<double>();
                WordedForecast = new List<string>();

                var periods = (JArray)jObject["properties"]["periods"];

                foreach (var period in periods)
                {
                    ForecastLabels.Add((string)period["name"]);
                    ForecastIcons.Add((string)period["icon"]);
                    ForecastDescriptions.Add((string)period["shortForecast"]);
                    WordedForecast.Add((string)period["detailedForecast"]);
                    double.TryParse((string)period["temperature"], out double temperature);
                    ForecastTemperatures.Add(temperature);
                }
            }
        }
    }
}

File: native/app/src/main/java/com/companyname/weather/services/NWSService.kt
package com.companyname.weather.services

import android.annotation.SuppressLint
import android.location.Location
import android.os.Handler
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.companyname.weather.services.nws.NWSGridPointsForecast
import com.companyname.weather.services.nws.NWSPoints
import com.companyname.weather.services.nws.NWSPointsStations
import com.companyname.weather.services.nws.NWSStationsObservations
import com.google.gson.Gson
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.lang.Exception
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class NWSService : LifecycleService() {

    companion object {
        var instance: NWSService? = null

        var location: MutableLiveData<String> = MutableLiveData()
        var conditions: MutableLiveData<Conditions> = MutableLiveData()
        var forecasts: MutableLiveData<List<Forecast>> = MutableLiveData()

        private var timestamp: Long = 0
    }

    data class Conditions(
        val dewPoint: Double? = null,
        val relativeHumidity: Double? = null,
        val temperature: Double? = null,
        val windDirection: Double? = null,
        val windSpeed: Double? = null,
        val windGust: Double? = null,
        val icon: String? = null,
        val textDescription: String? = null,
        val barometricPressure: Double? = null,
        val visibility: Double? = null,
        val windChill: Double? = null,
        val heatIndex: Double? = null,
        val timestamp: Date? = null
    )

    data class Forecast(
        val name: String? = null,
        val icon: String? = null,
        val shortForecast: String? = null,
        val detailedForecast: String? = null,
        val temperature: Double? = null,
        val windSpeed: String? = null,
        val windDirection: String? = null,
        val temperatureTrend: String? = null,
        val isDaytime: Boolean = false
    )

    private val refreshInterval: Long = 1000 * 60 * 15

    //    private val refreshInterval: Long = 1000 * 15
    private val mutex = Mutex()
    private var lastLocation: Location = Location("")
    private var handler: Handler? = null

    private var stationId: String? = null
    private var gridX: Int? = null
    private var gridY: Int? = null
    private var gridWFO: String? = null
    private var baseURL = "https://api.weather.gov"

    override fun onCreate() {
        super.onCreate()
        instance = this

        LocationService.location.observe(this, Observer { setLocation(it) })

        handler = Handler()
        var runnable = Runnable { }
        runnable = Runnable {
            GlobalScope.launch { refresh() }
            handler?.postDelayed(runnable, refreshInterval)
        }
        handler?.postDelayed(runnable, refreshInterval)
    }

    override fun onDestroy() {
        super.onDestroy()
        instance = null
        handler?.removeCallbacks(null)
    }

    private fun setLocation(location: Location) {
        if (lastLocation.distanceTo(location) > 100) {
            lastLocation = location
            stationId = null
            gridX = null
            gridY = null
            gridWFO = null
            timestamp = 0
            GlobalScope.launch { refresh() }
        }
    }

    private suspend fun refresh() {
        mutex.withLock {
            val duration = Date().time - timestamp
            if (duration > refreshInterval) {
                timestamp = Date().time
                if ((lastLocation.latitude == 0.0 && lastLocation.longitude == 0.0)) {
                    location.postValue(null)
                    conditions.postValue(null)
                    forecasts.postValue(null)
                } else {
                    getStation()
                    getConditions()
                    getForecast()
                }
            }
        }
    }

    private fun getStation() {
        try {
            val latitude = lastLocation.latitude
            val longitude = lastLocation.longitude
            val url = URL("$baseURL/points/$latitude,$longitude/stations")
            val response = url.readText()
            val points = Gson().fromJson(response, NWSPointsStations.Root::class.java)
            location.postValue(points.features[0].properties.name)
            stationId = points.features[0].properties.stationIdentifier
        } catch (ex: Exception) {
        }
    }

    @SuppressLint("SimpleDateFormat")
    private fun getConditions() {
        stationId ?: return

        try {
            val url = URL("$baseURL/stations/$stationId/observations/latest?require_qc=false")
            val response = url.readText()
            val stations = Gson().fromJson(response, NWSStationsObservations.Root::class.java)

            conditions.postValue(
                Conditions(
                    dewPoint = stations.properties.dewpoint.value,
                    relativeHumidity = stations.properties.relativeHumidity.value,
                    temperature = stations.properties.temperature.value,
                    windDirection = stations.properties.windDirection.value,
                    windSpeed = stations.properties.windSpeed.value,
                    windGust = stations.properties.windGust.value,
                    icon = stations.properties.icon,
                    textDescription = stations.properties.textDescription,
                    barometricPressure = stations.properties.barometricPressure.value,
                    visibility = stations.properties.visibility.value,
                    windChill = stations.properties.windChill.value,
                    heatIndex = stations.properties.heatIndex.value,
                    timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:sszzzzz").parse(stations.properties.timestamp)
                )
            )
        } catch (ex: Exception) {
        }
    }

    private fun getForecast() {
        if (listOf(gridX, gridY, gridWFO).any { it == null })
            getGridPoint()

        gridX ?: return
        gridY ?: return
        gridWFO ?: return

        try {
            val url = URL("$baseURL/gridpoints/$gridWFO/$gridX,$gridY/forecast")
            val response = url.readText()
            val forecast = Gson().fromJson(response, NWSGridPointsForecast.Root::class.java)

            forecasts.postValue(forecast.properties.periods.map {
                Forecast(
                    name = it.name,
                    icon = it.icon,
                    shortForecast = it.shortForecast,
                    detailedForecast = it.detailedForecast,
                    temperature = it.temperature,
                    windSpeed = it.windSpeed,
                    windDirection = it.windDirection,
                    temperatureTrend = it.temperatureTrend,
                    isDaytime = it.isDaytime
                )
            })
        } catch (ex: Exception) {
        }
    }

    private fun getGridPoint() {
        try {
            val latitude = lastLocation.latitude
            val longitude = lastLocation.longitude
            val url = URL("$baseURL/points/$latitude,$longitude")
            val response = url.readText()
            val points = Gson().fromJson(response, NWSPoints.Root::class.java)

            gridWFO = points.properties.cwa
            gridX = points.properties.gridX
            gridY = points.properties.gridY
        } catch (ex: Exception) {
        }
    }
}